master/buildbot/locks.py

```diff
@@ -206,31 +209,40 @@ def isOwner(self, owner, access):
         return (owner, access) in self.owners
 
 
-class RealMasterLock(BaseLock):
+class RealMasterLock(BaseLock, service.SharedService):
 
-    def __init__(self, lockid):
-        super().__init__(lockid.name, lockid.maxCount)
+    def __init__(self, name):
+        # the caller will want to call updateFromLockId after initialization most of the time.
+        super().__init__(name, 0)
+        self.config_version = 0
         self._updateDescription()
 
     def _updateDescription(self):
-        self.description = "<MasterLock({}, {})>".format(self.name,
+        self.description = "<MasterLock({}, {})>".format(self.lockName,
                                                          self.maxCount)
 
     def getLockForWorker(self, workername):
         return self
 
-    def updateFromLockId(self, lockid):
-        assert self.name == lockid.name
+    def updateFromLockId(self, lockid, config_version):
+        assert self.lockName == lockid.name
+        assert isinstance(config_version, int)
+
+        self.config_version = config_version
         self.setMaxCount(lockid.maxCount)
         self._updateDescription()
 
 
-class RealWorkerLock:
+class RealWorkerLock(service.SharedService):
 
-    def __init__(self, lockid):
-        self.name = lockid.name
-        self.maxCount = lockid.maxCount
-        self.maxCountForWorker = lockid.maxCountForWorker
+    def __init__(self, name):
+        super().__init__()
+
+        # the caller will want to call updateFromLockId after initialization most of the time
+        self.lockName = name
+        self.maxCount = None
+        self.maxCountForWorker = None
+        self.config_version = 0
```

Comment:
why not use self.config_version = -1 to force updateFromLockId at creation, without needing the new hasService api

Reply:
Fixed now.